// @syncfusion/ej2-angular-charts/esm2020/src/circularchart3d/selecteddataindexes.directive.mjs
import { Directive, ContentChildren } from '@angular/core';
import { ComplexBase, ArrayBase, setValue } from '@syncfusion/ej2-angular-base';
import * as i0 from "@angular/core";
let input = ['point', 'series'];
let outputs = [];
/**
 * Selected Data Directive
 * ```html
 * <e-selecteddataindexes><e-selecteddataindex></e-selecteddataindex><e-selecteddataindexes>
 * ```
 */
export class CircularChart3DSelectedDataIndexDirective extends ComplexBase {
    constructor(viewContainerRef) {
        super();
        this.viewContainerRef = viewContainerRef;
        setValue('currentInstance', this, this.viewContainerRef);
        this.registerEvents(outputs);
        this.directivePropList = input;
    }
}
CircularChart3DSelectedDataIndexDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: CircularChart3DSelectedDataIndexDirective, deps: [{ token: i0.ViewContainerRef }], target: i0.ɵɵFactoryTarget.Directive });
CircularChart3DSelectedDataIndexDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.0.3", type: CircularChart3DSelectedDataIndexDirective, selector: "e-circularchart3d-selecteddataindexes>e-circularchart3d-selecteddataindex", inputs: { point: "point", series: "series" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: CircularChart3DSelectedDataIndexDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'e-circularchart3d-selecteddataindexes>e-circularchart3d-selecteddataindex',
                    inputs: input,
                    outputs: outputs,
                    queries: {}
                }]
        }], ctorParameters: function () { return [{ type: i0.ViewContainerRef }]; } });
/**
 * CircularChart3DSelectedDataIndex Array Directive
 * @private
 */
export class CircularChart3DSelectedDataIndexesDirective extends ArrayBase {
    constructor() {
        super('selecteddataindexes');
    }
}
CircularChart3DSelectedDataIndexesDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: CircularChart3DSelectedDataIndexesDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
CircularChart3DSelectedDataIndexesDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.0.3", type: CircularChart3DSelectedDataIndexesDirective, selector: "ejs-circularchart3d>e-circularchart3d-selecteddataindexes", queries: [{ propertyName: "children", predicate: CircularChart3DSelectedDataIndexDirective }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: CircularChart3DSelectedDataIndexesDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ejs-circularchart3d>e-circularchart3d-selecteddataindexes',
                    queries: {
                        children: new ContentChildren(CircularChart3DSelectedDataIndexDirective)
                    },
                }]
        }], ctorParameters: function () { return []; } });
