// angular-password-strength-meter/esm2020/lib/password-strength-meter.service.mjs
import { Inject, Injectable, Optional } from '@angular/core';
import { zxcvbn, zxcvbnOptions } from '@zxcvbn-ts/core';
import { PSM_CONFIG } from './password-strength-meter.types';
import zxcvbnEnPackage from '@zxcvbn-ts/language-en';
import * as i0 from "@angular/core";
export class IPasswordStrengthMeterService {
}
export const DEFAULT_CONFIG = {
    translations: zxcvbnEnPackage.translations,
};
export class PasswordStrengthMeterService extends IPasswordStrengthMeterService {
    constructor(options) {
        super();
        if (options) {
            zxcvbnOptions.setOptions(options);
        }
        else {
            zxcvbnOptions.setOptions(DEFAULT_CONFIG);
        }
    }
    /**
     *  this will return the password strength score in number
     *  0 - too guessable
     *  1 - very guessable
     *  2 - somewhat guessable
     *  3 - safely unguessable
     *  4 - very unguessable
     *
     *  @param password - Password
     */
    score(password) {
        const result = zxcvbn(password);
        return result.score;
    }
    /**
     * this will return the password strength score with feedback messages
     * return type { score: number; feedback: { suggestions: string[]; warning: string } }
     *
     * @param password - Password
     */
    scoreWithFeedback(password) {
        const result = zxcvbn(password);
        return { score: result.score, feedback: result.feedback };
    }
}
PasswordStrengthMeterService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterService, deps: [{ token: PSM_CONFIG, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
PasswordStrengthMeterService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [PSM_CONFIG]
                }] }]; } });
