// angular-password-strength-meter/esm2020/lib/password-strength-meter.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PasswordStrengthMeterComponent } from './password-strength-meter.component';
import { PSMProgressBarDirective } from './psm-progress-bar.directive';
import { IPasswordStrengthMeterService, PasswordStrengthMeterService, } from './password-strength-meter.service';
import * as i0 from "@angular/core";
const defaultOptions = {
    serviceClass: PasswordStrengthMeterService,
};
export class PasswordStrengthMeterModule {
    static forRoot({ serviceClass, } = defaultOptions) {
        return {
            ngModule: PasswordStrengthMeterModule,
            providers: [
                {
                    provide: IPasswordStrengthMeterService,
                    useClass: serviceClass,
                },
            ],
        };
    }
    static forChild() {
        return {
            ngModule: PasswordStrengthMeterModule,
        };
    }
}
PasswordStrengthMeterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PasswordStrengthMeterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterModule, declarations: [PasswordStrengthMeterComponent, PSMProgressBarDirective], imports: [CommonModule], exports: [PasswordStrengthMeterComponent, PSMProgressBarDirective] });
PasswordStrengthMeterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [PasswordStrengthMeterComponent, PSMProgressBarDirective],
                    exports: [PasswordStrengthMeterComponent, PSMProgressBarDirective],
                }]
        }] });
