// angular-password-strength-meter/esm2020/lib/password-strength-meter.component.mjs
import { Component, Input, Output, EventEmitter, HostBinding, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./password-strength-meter.service";
import * as i2 from "@angular/common";
import * as i3 from "./psm-progress-bar.directive";
export class PasswordStrengthMeterComponent {
    constructor(passwordStrengthMeterService) {
        this.passwordStrengthMeterService = passwordStrengthMeterService;
        this.minPasswordLength = 8;
        this.enableFeedback = false;
        this.colors = [];
        this.numberOfProgressBarItems = 5;
        this.strengthChange = new EventEmitter();
        this.baseClass = 'psm';
        this.passwordStrength = null;
        this.feedback = null;
        this.prevPasswordStrength = null;
    }
    ngOnChanges(changes) {
        if (changes.password) {
            this.calculatePasswordStrength();
        }
    }
    calculatePasswordStrength() {
        // TODO validation logic optimization
        if (!this.password) {
            this.passwordStrength = null;
            this.feedback = null;
        }
        else if (this.password && this.password.length < this.minPasswordLength) {
            this.passwordStrength = 0;
            this.feedback = null;
        }
        else {
            if (this.enableFeedback) {
                const result = this.passwordStrengthMeterService.scoreWithFeedback(this.password);
                this.passwordStrength = result.score;
                this.feedback = result.feedback;
            }
            else {
                this.passwordStrength = this.passwordStrengthMeterService.score(this.password);
                this.feedback = null;
            }
        }
        // Only emit the passwordStrength if it changed
        if (this.prevPasswordStrength !== this.passwordStrength) {
            this.strengthChange.emit(this.passwordStrength);
            this.prevPasswordStrength = this.passwordStrength;
        }
    }
}
PasswordStrengthMeterComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterComponent, deps: [{ token: i1.IPasswordStrengthMeterService }], target: i0.ɵɵFactoryTarget.Component });
PasswordStrengthMeterComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.9", type: PasswordStrengthMeterComponent, selector: "password-strength-meter", inputs: { password: "password", minPasswordLength: "minPasswordLength", enableFeedback: "enableFeedback", colors: "colors", numberOfProgressBarItems: "numberOfProgressBarItems" }, outputs: { strengthChange: "strengthChange" }, host: { properties: { "class": "this.baseClass" } }, usesOnChanges: true, ngImport: i0, template: "<div\n  class=\"psm__progress-bar\"\n  role=\"progressbar\"\n  [passwordStrength]=\"passwordStrength\"\n  [numberOfProgressBarItems]=\"numberOfProgressBarItems\"\n  [colors]=\"colors\"\n>\n  <div class=\"psm__progress-bar-items\"></div>\n  <div class=\"psm__progress-bar-overlay\"></div>\n</div>\n<ng-container *ngIf=\"enableFeedback && feedback\">\n  <small class=\"psm__feedback\" *ngIf=\"feedback.warning\">\n    {{ feedback.warning }}\n  </small>\n  <small\n    class=\"psm__suggestion\"\n    *ngIf=\"feedback.suggestions && feedback.suggestions.length\"\n  >\n    <ng-container\n      *ngFor=\"let suggestion of feedback.suggestions; last as isLast\"\n    >\n      {{ suggestion }}{{ isLast ? '' : ' ' }}\n    </ng-container>\n  </small>\n</ng-container>\n", styles: ["*,*:before,*:after{box-sizing:border-box}.psm__progress-bar{position:relative;height:3px;margin:10px auto;border-radius:3px;background:#fff}.psm__progress-bar-items{display:flex;justify-content:space-evenly;width:100%;height:100%;background:#ddd}.psm__progress-bar-item{height:100%;z-index:5;border-right:4px solid #fff}.psm__progress-bar-item:last-child{border-right:0px}.psm__progress-bar-overlay{position:absolute;background:red;border-radius:3px;height:3px;top:0;transition:width .5s ease-in-out,background .25s}.psm__feedback,.psm__suggestion{font-size:70%;font-weight:400;color:#6c757d!important;display:block;margin-top:.25rem}\n"], dependencies: [{ kind: "directive", type: i2.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i3.PSMProgressBarDirective, selector: ".psm__progress-bar", inputs: ["numberOfProgressBarItems", "passwordStrength", "colors"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterComponent, decorators: [{
            type: Component,
            args: [{ selector: 'password-strength-meter', template: "<div\n  class=\"psm__progress-bar\"\n  role=\"progressbar\"\n  [passwordStrength]=\"passwordStrength\"\n  [numberOfProgressBarItems]=\"numberOfProgressBarItems\"\n  [colors]=\"colors\"\n>\n  <div class=\"psm__progress-bar-items\"></div>\n  <div class=\"psm__progress-bar-overlay\"></div>\n</div>\n<ng-container *ngIf=\"enableFeedback && feedback\">\n  <small class=\"psm__feedback\" *ngIf=\"feedback.warning\">\n    {{ feedback.warning }}\n  </small>\n  <small\n    class=\"psm__suggestion\"\n    *ngIf=\"feedback.suggestions && feedback.suggestions.length\"\n  >\n    <ng-container\n      *ngFor=\"let suggestion of feedback.suggestions; last as isLast\"\n    >\n      {{ suggestion }}{{ isLast ? '' : ' ' }}\n    </ng-container>\n  </small>\n</ng-container>\n", styles: ["*,*:before,*:after{box-sizing:border-box}.psm__progress-bar{position:relative;height:3px;margin:10px auto;border-radius:3px;background:#fff}.psm__progress-bar-items{display:flex;justify-content:space-evenly;width:100%;height:100%;background:#ddd}.psm__progress-bar-item{height:100%;z-index:5;border-right:4px solid #fff}.psm__progress-bar-item:last-child{border-right:0px}.psm__progress-bar-overlay{position:absolute;background:red;border-radius:3px;height:3px;top:0;transition:width .5s ease-in-out,background .25s}.psm__feedback,.psm__suggestion{font-size:70%;font-weight:400;color:#6c757d!important;display:block;margin-top:.25rem}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.IPasswordStrengthMeterService }]; }, propDecorators: { password: [{
                type: Input
            }], minPasswordLength: [{
                type: Input
            }], enableFeedback: [{
                type: Input
            }], colors: [{
                type: Input
            }], numberOfProgressBarItems: [{
                type: Input
            }], strengthChange: [{
                type: Output
            }], baseClass: [{
                type: HostBinding,
                args: ['class']
            }] } });
