// angular-password-strength-meter/esm2020/lib/psm-progress-bar.directive.mjs
import { Directive, HostBinding, Input, } from '@angular/core';
import * as i0 from "@angular/core";
export class PSMProgressBarDirective {
    constructor(renderer, el) {
        this.renderer = renderer;
        this.el = el;
        this.colors = [];
        this.minProgressVal = 0;
        this.maxProgressVal = 100;
        this.currentProgressVal = 0;
        this.dataPasswordStrength = 0;
        this.defaultColors = [
            'darkred',
            'orangered',
            'orange',
            'yellowgreen',
            'green',
        ];
        this.progressBar = this.el.nativeElement;
    }
    ngOnChanges(changes) {
        if (changes.numberOfProgressBarItems) {
            this.setProgressBarItems();
        }
        this.setProgressBar();
    }
    setProgressBarItems() {
        const progressBarItemContainer = this.progressBar.querySelector('.psm__progress-bar-items');
        const width = 100 / this.numberOfProgressBarItems;
        progressBarItemContainer.childNodes.forEach((item) => {
            this.renderer.removeChild(progressBarItemContainer, item);
        });
        Array(this.numberOfProgressBarItems)
            .fill(1)
            .forEach((_) => {
            const progressBarItem = this.renderer.createElement('div');
            this.renderer.addClass(progressBarItem, 'psm__progress-bar-item');
            this.renderer.setStyle(progressBarItem, 'width', `${width}%`);
            this.renderer.appendChild(progressBarItemContainer, progressBarItem);
        });
    }
    setProgressBar() {
        const progressBarOverlayWidth = this.getFillMeterWidth(this.passwordStrength);
        const progressBarOverlayWidthInPx = `${progressBarOverlayWidth}%`;
        const progressLevelBasedOnItems = (progressBarOverlayWidth / 100) * this.numberOfProgressBarItems;
        const progressBarOverlayColor = this.getMeterFillColor(progressLevelBasedOnItems);
        this.dataPasswordStrength = this.passwordStrength || 0;
        this.currentProgressVal = progressBarOverlayWidth;
        const overlayElement = this.progressBar.querySelector('.psm__progress-bar-overlay');
        if (overlayElement) {
            this.renderer.setStyle(overlayElement, 'width', progressBarOverlayWidthInPx);
            this.renderer.setStyle(overlayElement, 'background-color', progressBarOverlayColor);
        }
    }
    getFillMeterWidth(strength) {
        if (strength === null || strength === undefined) {
            return 0;
        }
        const strengthInPercentage = strength !== null ? ((strength + 1) / 5) * 100 : 0;
        const roundedStrengthInPercentage = this.getRoundedStrength(strengthInPercentage, 100 / this.numberOfProgressBarItems);
        return roundedStrengthInPercentage;
    }
    getMeterFillColor(progressLevel) {
        if (!progressLevel ||
            progressLevel <= 0 ||
            (progressLevel > this.colors.length &&
                progressLevel > this.defaultColors.length)) {
            return this.colors[0] ? this.colors[0] : this.defaultColors[0];
        }
        const index = progressLevel - 1;
        return this.colors[index] ? this.colors[index] : this.defaultColors[index];
    }
    getRoundedStrength(strength, roundTo) {
        return Math.round(strength / roundTo) * roundTo;
    }
}
PSMProgressBarDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PSMProgressBarDirective, deps: [{ token: i0.Renderer2 }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
PSMProgressBarDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.9", type: PSMProgressBarDirective, selector: ".psm__progress-bar", inputs: { numberOfProgressBarItems: "numberOfProgressBarItems", passwordStrength: "passwordStrength", colors: "colors" }, host: { properties: { "attr.aria-valuemin": "this.minProgressVal", "attr.aria-valuemax": "this.maxProgressVal", "attr.aria-valuenow": "this.currentProgressVal", "attr.data-strength": "this.dataPasswordStrength" } }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PSMProgressBarDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line
                    selector: '.psm__progress-bar',
                }]
        }], ctorParameters: function () { return [{ type: i0.Renderer2 }, { type: i0.ElementRef }]; }, propDecorators: { numberOfProgressBarItems: [{
                type: Input
            }], passwordStrength: [{
                type: Input
            }], colors: [{
                type: Input
            }], minProgressVal: [{
                type: HostBinding,
                args: ['attr.aria-valuemin']
            }], maxProgressVal: [{
                type: HostBinding,
                args: ['attr.aria-valuemax']
            }], currentProgressVal: [{
                type: HostBinding,
                args: ['attr.aria-valuenow']
            }], dataPasswordStrength: [{
                type: HostBinding,
                args: ['attr.data-strength']
            }] } });
