// angular-password-strength-meter/fesm2020/angular-password-strength-meter.mjs
import * as i0 from '@angular/core';
import { InjectionToken, Injectable, Optional, Inject, Directive, Input, HostBinding, EventEmitter, Component, Output, NgModule } from '@angular/core';
import { zxcvbnOptions, zxcvbn } from '@zxcvbn-ts/core';
import zxcvbnEnPackage from '@zxcvbn-ts/language-en';
import * as i2 from '@angular/common';
import { CommonModule } from '@angular/common';

const PSM_CONFIG = new InjectionToken('PSM_CONFIG');

class IPasswordStrengthMeterService {
}
const DEFAULT_CONFIG = {
    translations: zxcvbnEnPackage.translations,
};
class PasswordStrengthMeterService extends IPasswordStrengthMeterService {
    constructor(options) {
        super();
        if (options) {
            zxcvbnOptions.setOptions(options);
        }
        else {
            zxcvbnOptions.setOptions(DEFAULT_CONFIG);
        }
    }
    /**
     *  this will return the password strength score in number
     *  0 - too guessable
     *  1 - very guessable
     *  2 - somewhat guessable
     *  3 - safely unguessable
     *  4 - very unguessable
     *
     *  @param password - Password
     */
    score(password) {
        const result = zxcvbn(password);
        return result.score;
    }
    /**
     * this will return the password strength score with feedback messages
     * return type { score: number; feedback: { suggestions: string[]; warning: string } }
     *
     * @param password - Password
     */
    scoreWithFeedback(password) {
        const result = zxcvbn(password);
        return { score: result.score, feedback: result.feedback };
    }
}
PasswordStrengthMeterService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterService, deps: [{ token: PSM_CONFIG, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
PasswordStrengthMeterService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [PSM_CONFIG]
                }] }]; } });

class PSMProgressBarDirective {
    constructor(renderer, el) {
        this.renderer = renderer;
        this.el = el;
        this.colors = [];
        this.minProgressVal = 0;
        this.maxProgressVal = 100;
        this.currentProgressVal = 0;
        this.dataPasswordStrength = 0;
        this.defaultColors = [
            'darkred',
            'orangered',
            'orange',
            'yellowgreen',
            'green',
        ];
        this.progressBar = this.el.nativeElement;
    }
    ngOnChanges(changes) {
        if (changes.numberOfProgressBarItems) {
            this.setProgressBarItems();
        }
        this.setProgressBar();
    }
    setProgressBarItems() {
        const progressBarItemContainer = this.progressBar.querySelector('.psm__progress-bar-items');
        const width = 100 / this.numberOfProgressBarItems;
        progressBarItemContainer.childNodes.forEach((item) => {
            this.renderer.removeChild(progressBarItemContainer, item);
        });
        Array(this.numberOfProgressBarItems)
            .fill(1)
            .forEach((_) => {
            const progressBarItem = this.renderer.createElement('div');
            this.renderer.addClass(progressBarItem, 'psm__progress-bar-item');
            this.renderer.setStyle(progressBarItem, 'width', `${width}%`);
            this.renderer.appendChild(progressBarItemContainer, progressBarItem);
        });
    }
    setProgressBar() {
        const progressBarOverlayWidth = this.getFillMeterWidth(this.passwordStrength);
        const progressBarOverlayWidthInPx = `${progressBarOverlayWidth}%`;
        const progressLevelBasedOnItems = (progressBarOverlayWidth / 100) * this.numberOfProgressBarItems;
        const progressBarOverlayColor = this.getMeterFillColor(progressLevelBasedOnItems);
        this.dataPasswordStrength = this.passwordStrength || 0;
        this.currentProgressVal = progressBarOverlayWidth;
        const overlayElement = this.progressBar.querySelector('.psm__progress-bar-overlay');
        if (overlayElement) {
            this.renderer.setStyle(overlayElement, 'width', progressBarOverlayWidthInPx);
            this.renderer.setStyle(overlayElement, 'background-color', progressBarOverlayColor);
        }
    }
    getFillMeterWidth(strength) {
        if (strength === null || strength === undefined) {
            return 0;
        }
        const strengthInPercentage = strength !== null ? ((strength + 1) / 5) * 100 : 0;
        const roundedStrengthInPercentage = this.getRoundedStrength(strengthInPercentage, 100 / this.numberOfProgressBarItems);
        return roundedStrengthInPercentage;
    }
    getMeterFillColor(progressLevel) {
        if (!progressLevel ||
            progressLevel <= 0 ||
            (progressLevel > this.colors.length &&
                progressLevel > this.defaultColors.length)) {
            return this.colors[0] ? this.colors[0] : this.defaultColors[0];
        }
        const index = progressLevel - 1;
        return this.colors[index] ? this.colors[index] : this.defaultColors[index];
    }
    getRoundedStrength(strength, roundTo) {
        return Math.round(strength / roundTo) * roundTo;
    }
}
PSMProgressBarDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PSMProgressBarDirective, deps: [{ token: i0.Renderer2 }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
PSMProgressBarDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.9", type: PSMProgressBarDirective, selector: ".psm__progress-bar", inputs: { numberOfProgressBarItems: "numberOfProgressBarItems", passwordStrength: "passwordStrength", colors: "colors" }, host: { properties: { "attr.aria-valuemin": "this.minProgressVal", "attr.aria-valuemax": "this.maxProgressVal", "attr.aria-valuenow": "this.currentProgressVal", "attr.data-strength": "this.dataPasswordStrength" } }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PSMProgressBarDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line
                    selector: '.psm__progress-bar',
                }]
        }], ctorParameters: function () { return [{ type: i0.Renderer2 }, { type: i0.ElementRef }]; }, propDecorators: { numberOfProgressBarItems: [{
                type: Input
            }], passwordStrength: [{
                type: Input
            }], colors: [{
                type: Input
            }], minProgressVal: [{
                type: HostBinding,
                args: ['attr.aria-valuemin']
            }], maxProgressVal: [{
                type: HostBinding,
                args: ['attr.aria-valuemax']
            }], currentProgressVal: [{
                type: HostBinding,
                args: ['attr.aria-valuenow']
            }], dataPasswordStrength: [{
                type: HostBinding,
                args: ['attr.data-strength']
            }] } });

class PasswordStrengthMeterComponent {
    constructor(passwordStrengthMeterService) {
        this.passwordStrengthMeterService = passwordStrengthMeterService;
        this.minPasswordLength = 8;
        this.enableFeedback = false;
        this.colors = [];
        this.numberOfProgressBarItems = 5;
        this.strengthChange = new EventEmitter();
        this.baseClass = 'psm';
        this.passwordStrength = null;
        this.feedback = null;
        this.prevPasswordStrength = null;
    }
    ngOnChanges(changes) {
        if (changes.password) {
            this.calculatePasswordStrength();
        }
    }
    calculatePasswordStrength() {
        // TODO validation logic optimization
        if (!this.password) {
            this.passwordStrength = null;
            this.feedback = null;
        }
        else if (this.password && this.password.length < this.minPasswordLength) {
            this.passwordStrength = 0;
            this.feedback = null;
        }
        else {
            if (this.enableFeedback) {
                const result = this.passwordStrengthMeterService.scoreWithFeedback(this.password);
                this.passwordStrength = result.score;
                this.feedback = result.feedback;
            }
            else {
                this.passwordStrength = this.passwordStrengthMeterService.score(this.password);
                this.feedback = null;
            }
        }
        // Only emit the passwordStrength if it changed
        if (this.prevPasswordStrength !== this.passwordStrength) {
            this.strengthChange.emit(this.passwordStrength);
            this.prevPasswordStrength = this.passwordStrength;
        }
    }
}
PasswordStrengthMeterComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterComponent, deps: [{ token: IPasswordStrengthMeterService }], target: i0.ɵɵFactoryTarget.Component });
PasswordStrengthMeterComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.9", type: PasswordStrengthMeterComponent, selector: "password-strength-meter", inputs: { password: "password", minPasswordLength: "minPasswordLength", enableFeedback: "enableFeedback", colors: "colors", numberOfProgressBarItems: "numberOfProgressBarItems" }, outputs: { strengthChange: "strengthChange" }, host: { properties: { "class": "this.baseClass" } }, usesOnChanges: true, ngImport: i0, template: "<div\n  class=\"psm__progress-bar\"\n  role=\"progressbar\"\n  [passwordStrength]=\"passwordStrength\"\n  [numberOfProgressBarItems]=\"numberOfProgressBarItems\"\n  [colors]=\"colors\"\n>\n  <div class=\"psm__progress-bar-items\"></div>\n  <div class=\"psm__progress-bar-overlay\"></div>\n</div>\n<ng-container *ngIf=\"enableFeedback && feedback\">\n  <small class=\"psm__feedback\" *ngIf=\"feedback.warning\">\n    {{ feedback.warning }}\n  </small>\n  <small\n    class=\"psm__suggestion\"\n    *ngIf=\"feedback.suggestions && feedback.suggestions.length\"\n  >\n    <ng-container\n      *ngFor=\"let suggestion of feedback.suggestions; last as isLast\"\n    >\n      {{ suggestion }}{{ isLast ? '' : ' ' }}\n    </ng-container>\n  </small>\n</ng-container>\n", styles: ["*,*:before,*:after{box-sizing:border-box}.psm__progress-bar{position:relative;height:3px;margin:10px auto;border-radius:3px;background:#fff}.psm__progress-bar-items{display:flex;justify-content:space-evenly;width:100%;height:100%;background:#ddd}.psm__progress-bar-item{height:100%;z-index:5;border-right:4px solid #fff}.psm__progress-bar-item:last-child{border-right:0px}.psm__progress-bar-overlay{position:absolute;background:red;border-radius:3px;height:3px;top:0;transition:width .5s ease-in-out,background .25s}.psm__feedback,.psm__suggestion{font-size:70%;font-weight:400;color:#6c757d!important;display:block;margin-top:.25rem}\n"], dependencies: [{ kind: "directive", type: i2.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: PSMProgressBarDirective, selector: ".psm__progress-bar", inputs: ["numberOfProgressBarItems", "passwordStrength", "colors"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterComponent, decorators: [{
            type: Component,
            args: [{ selector: 'password-strength-meter', template: "<div\n  class=\"psm__progress-bar\"\n  role=\"progressbar\"\n  [passwordStrength]=\"passwordStrength\"\n  [numberOfProgressBarItems]=\"numberOfProgressBarItems\"\n  [colors]=\"colors\"\n>\n  <div class=\"psm__progress-bar-items\"></div>\n  <div class=\"psm__progress-bar-overlay\"></div>\n</div>\n<ng-container *ngIf=\"enableFeedback && feedback\">\n  <small class=\"psm__feedback\" *ngIf=\"feedback.warning\">\n    {{ feedback.warning }}\n  </small>\n  <small\n    class=\"psm__suggestion\"\n    *ngIf=\"feedback.suggestions && feedback.suggestions.length\"\n  >\n    <ng-container\n      *ngFor=\"let suggestion of feedback.suggestions; last as isLast\"\n    >\n      {{ suggestion }}{{ isLast ? '' : ' ' }}\n    </ng-container>\n  </small>\n</ng-container>\n", styles: ["*,*:before,*:after{box-sizing:border-box}.psm__progress-bar{position:relative;height:3px;margin:10px auto;border-radius:3px;background:#fff}.psm__progress-bar-items{display:flex;justify-content:space-evenly;width:100%;height:100%;background:#ddd}.psm__progress-bar-item{height:100%;z-index:5;border-right:4px solid #fff}.psm__progress-bar-item:last-child{border-right:0px}.psm__progress-bar-overlay{position:absolute;background:red;border-radius:3px;height:3px;top:0;transition:width .5s ease-in-out,background .25s}.psm__feedback,.psm__suggestion{font-size:70%;font-weight:400;color:#6c757d!important;display:block;margin-top:.25rem}\n"] }]
        }], ctorParameters: function () { return [{ type: IPasswordStrengthMeterService }]; }, propDecorators: { password: [{
                type: Input
            }], minPasswordLength: [{
                type: Input
            }], enableFeedback: [{
                type: Input
            }], colors: [{
                type: Input
            }], numberOfProgressBarItems: [{
                type: Input
            }], strengthChange: [{
                type: Output
            }], baseClass: [{
                type: HostBinding,
                args: ['class']
            }] } });

const defaultOptions = {
    serviceClass: PasswordStrengthMeterService,
};
class PasswordStrengthMeterModule {
    static forRoot({ serviceClass, } = defaultOptions) {
        return {
            ngModule: PasswordStrengthMeterModule,
            providers: [
                {
                    provide: IPasswordStrengthMeterService,
                    useClass: serviceClass,
                },
            ],
        };
    }
    static forChild() {
        return {
            ngModule: PasswordStrengthMeterModule,
        };
    }
}
PasswordStrengthMeterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PasswordStrengthMeterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterModule, declarations: [PasswordStrengthMeterComponent, PSMProgressBarDirective], imports: [CommonModule], exports: [PasswordStrengthMeterComponent, PSMProgressBarDirective] });
PasswordStrengthMeterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: PasswordStrengthMeterModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [PasswordStrengthMeterComponent, PSMProgressBarDirective],
                    exports: [PasswordStrengthMeterComponent, PSMProgressBarDirective],
                }]
        }] });

/*
 * Public API Surface of password-strength-meter
 */

/**
 * Generated bundle index. Do not edit.
 */

export { DEFAULT_CONFIG, IPasswordStrengthMeterService, PSMProgressBarDirective, PSM_CONFIG, PasswordStrengthMeterComponent, PasswordStrengthMeterModule, PasswordStrengthMeterService };
